### PR TITLE
Respawn API changes

### PR DESCRIPTION
**Pull request by Mikihero**
## Description
**Describe the changes** 
Added and changed some parts of the respawn api to reflect new changes, not done by any means

**What is the current behavior?** (You can also link to an open issue here)

**What is the new behavior?** (if this is a feature change)

**Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

**Other information**:

<br />

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentations
<br />

## Submission checklist

- [ ] I have checked the project can be compiled
- [ ] I have tested my changes and it worked as expected

### Patches (if there are any changes related to Harmony patches)
- [ ] I have checked no IL patching errors in the console

### Other
- [ ] Still requires more testing

### CODE REVIEW

**Comment by Mikihero** [comment]
a lot of this is adapted from `WaveManager`, still need to finish it and write & update docs

File: EXILED/Exiled.API/Features/Respawn.cs [hunk 1]
```diff
@@ -272,7 +261,7 @@ public static void ModifyTickets(Faction team, int amount)
         /// </summary>
         /// <param name="faction"><see cref="SpawnableTeamType"/>'s faction.</param>
         /// <returns>Tickets of team or <c>-1</c> if team doesn't depend on tickets.</returns>
-        public static int GetTickets(SpawnableFaction faction)
+        public static int GetTokens(SpawnableFaction faction)
```
Comment (Misaka-ZeroTwo) [inline EXILED/Exiled.API/Features/Respawn.cs:264]: Why introduce a break change here?

File: EXILED/Exiled.API/Features/Respawn.cs [hunk 2]
```diff
@@ -283,11 +272,12 @@ public static int GetTickets(SpawnableFaction faction)
...
         /// </summary>
-        /// <param name="team">The <see cref="SpawnableTeamType"/> to spawn.</param>
-        public static void ForceWave(Faction team)
+        /// <param name="faction">The <see cref="SpawnableTeamType"/> to spawn.</param>
+        /// <param name="mini">Docs.</param>
+        public static void ForceWave(Faction faction, bool mini = false)
         {
-            if (TryGetWaveBase(team, out SpawnableWaveBase wave))
-                ForceWave(wave);
+            SpawnableWaveBase waveBase = WaveManager.Waves.Find(x => x.TargetFaction == faction && (mini && x is IMiniWave));
```
Comment (Misaka-ZeroTwo) [inline EXILED/Exiled.API/Features/Respawn.cs:279]:
```suggestion
            SpawnableWaveBase waveBase = WaveManager.Waves.Find(x => x.TargetFaction == faction && (mini == x is IMiniWave));
```

Comment (Misaka-ZeroTwo) [reply EXILED/Exiled.API/Features/Respawn.cs:279]: I hope I understand that part correctly.

Comment (Mikihero) [reply EXILED/Exiled.API/Features/Respawn.cs:279]: what I had was correct, `mini` is the parameter that determines whether the user wants a minu wave or a normal wave, all 4 waves (2 mini waves and 2 normal ones) are stored in `WaveManager.Waves`, the `x` here is an instance of a wave

File: EXILED/Exiled.API/Features/Respawn.cs (hunk 1 above)
Comment (Mikihero) [reply EXILED/Exiled.API/Features/Respawn.cs:264]: Everything related to respawning will have to be reworked anyway so I thought we might as well name it correctly to not cause confusion

Comment (Misaka-ZeroTwo) [reply EXILED/Exiled.API/Features/Respawn.cs:264]: Ummm, I think it's best to deprecate the old one and add a new one. It's not allowed nor needed to make break changes when unnecessary.

Comment (Mikihero) [reply EXILED/Exiled.API/Features/Respawn.cs:264]: actually this isn't a breaking change, this method was added here https://github.com/ExMod-Team/EXILED/commit/07113063bfa0ea99f2a2bbb909c3a5e39cdb8115#diff-446a27f4f248ca3b4c61f854e9c442c310dc859c4658c93a4d4145b897b631edR263-R269

**Comment by Mikihero** [comment]
there are 100% some things I shouldn't have done or should've done differently but I'm too tired to go through them myself, @Misaka-ZeroTwo @Misfiy if you could take another look and point everything out that'd be great

File: EXILED/Exiled.API/Features/Respawn.cs [hunk 3]
```diff
@@ -231,49 +267,79 @@ public static void SummonNtfChopper()
...
         /// <param name="amount">The amount of tickets to grant.</param>
-        public static void GrantTickets(Faction team, int amount)
+        /// <returns>Success.</returns>
+        public static bool GrantTokens(Faction faction, int amount)
         {
-            if (TryGetWaveBase(team, out SpawnableWaveBase wave) && wave is ILimitedWave limitedWave)
-                limitedWave.RespawnTokens += amount;
+            if (TryGetWaveBases(faction, out IEnumerable<SpawnableWaveBase> waveBases))
+            {
+                foreach (ILimitedWave limitedWave in waveBases.Select(x => (ILimitedWave)x))
```
Comment (VALERA771) [inline EXILED/Exiled.API/Features/Respawn.cs:284]:
```suggestion
                foreach (ILimitedWave limitedWave in waveBases.OfType<ILimitedWave>())
```
This should work. Yeah, currently we don't have waves that don't inherit this interface but... idk why nw has added it. I thimk there was some point

File: EXILED/Exiled.API/Features/Respawn.cs [hunk 4]
```diff
@@ -231,49 +267,79 @@ public static void SummonNtfChopper()
...
         /// <param name="amount">The amount of tickets to remove.</param>
-        public static void RemoveTickets(Faction team, int amount)
+        /// <returns>Success.</returns>
+        public static bool RemoveTokens(Faction faction, int amount)
         {
-            if (TryGetWaveBase(team, out SpawnableWaveBase wave) && wave is ILimitedWave limitedWave)
-                limitedWave.RespawnTokens = Math.Max(0, limitedWave.RespawnTokens - amount);
+            if (TryGetWaveBases(faction, out IEnumerable<SpawnableWaveBase> waveBases))
+            {
+                foreach (ILimitedWave limitedWave in waveBases.Select(x => (ILimitedWave)x))
```
Comment (VALERA771) [inline EXILED/Exiled.API/Features/Respawn.cs:305]: Same as above

File: EXILED/Exiled.API/Features/Respawn.cs [hunk 5]
```diff
@@ -231,49 +267,79 @@ public static void SummonNtfChopper()
...
         /// <param name="amount">The amount of tickets to modify.</param>
-        public static void ModifyTickets(Faction team, int amount)
+        /// <returns>Success.</returns>
+        public static bool ModifyTokens(Faction faction, int amount)
         {
-            if (TryGetWaveBase(team, out SpawnableWaveBase wave) && wave is ILimitedWave limitedWave)
-                limitedWave.RespawnTokens = amount;
+            if (TryGetWaveBases(faction, out IEnumerable<SpawnableWaveBase> waveBases))
+            {
+                foreach (ILimitedWave limitedWave in waveBases.Select(x => (ILimitedWave)x))
```
Comment (VALERA771) [inline EXILED/Exiled.API/Features/Respawn.cs:326]: Same as above

File: EXILED/Exiled.API/Extensions/RoleExtensions.cs [hunk 6]
```diff
@@ -216,12 +217,61 @@ public static Dictionary<AmmoType, ushort> GetStartingAmmo(this RoleTypeId roleT
...
+        /// Docs1.
+        /// </summary>
+        /// <param name="spawnableFaction">Docs1..</param>
+        /// <returns>Docs2.</returns>
+        public static Faction GetFaction(this SpawnableFaction spawnableFaction) => spawnableFaction switch
+        {
+            SpawnableFaction.ChaosWave => Faction.FoundationEnemy,
+            SpawnableFaction.ChaosMiniWave => Faction.FoundationEnemy,
+            SpawnableFaction.NtfWave => Faction.FoundationStaff,
+            _ => Faction.FoundationStaff
```
[the remark this answers is not in the document]
Comment (Mikihero) [reply EXILED/Exiled.API/Extensions/RoleExtensions.cs:238]: wasn't sure if that'd work

**Comment by Mikihero** [comment]
> You can argue with me about this comments but imho this is just better

agree on all of those lmao

**Comment by Mikihero** [comment]
@VALERA771 do you think I should add more overloads to methods like `ForceWave` and similar to accept all the other ways of representing a spawn wave? shit like `SpawnableTeamType` `Faction` `SpawnableFaction`. Right now you can do it by using extension methods that I've added but I could just add overloads that do that for you

**Comment by Mikihero** [comment]
I'm too tired too think clearly rn, someone tell me if I overcooked with this last addition

**Comment by VALERA771** [comment]
Start writing docs pls. And what overloads on spawning method you want to add?

**Comment by Mikihero** [comment]
> Start writing docs pls. And what overloads on spawning method you want to add?

I'll get a good chunks of the docs done today. I was talking about methods like `RemoveTokens` `ModifyTokens` `GrantTokens` and `PlayEffect`

**Comment by Mikihero** [comment]
pls tell me if I've missed something

**Comment by VALERA771** [comment]
You've missed writing docs 😄

**Comment by Mikihero** [comment]
wait I forgot docs for rolextensions

**Comment by VALERA771** [comment]
I don't think that we need so many seealso on overloads

**Comment by Mikihero** [comment]
> I don't think that we need so many seealso on overloads

I guess that makes sense, I'll just add 1 for different (but similar) methods instead of for every overload

**Comment by VALERA771** [comment]
Timer methods have too much seealso xml docs

**Review by VALERA771 — APPROVED** [review]

**Review by louis1706 — APPROVED** [review]

**Pull request merged**